pkg/controller/controller_instance.go

```diff
@@ -70,6 +70,113 @@ func (c *controller) instanceUpdate(oldObj, newObj interface{}) {
+	// this can happen if the service class name referenced never existed.
+	if !instance.Status.AsyncOpInProgress && instance.Status.Checksum == nil {
+		finalizers.Delete(finalizerToken)
+		// Clear the finalizer
+		return c.updateInstanceFinalizers(instance, finalizers.List())
+	}
+
+	// All updates not having a DeletingTimestamp will have been handled above
+	// and returned early. If we reach this point, we're dealing with an update
+	// that's actually a soft delete-- i.e. we have some finalization to do.
+	// Since the potential exists for an instance to have multiple finalizers and
+	// since those most be cleared in order, we proceed with the soft delete
+	// only if it's "our turn--" i.e. only if the finalizer we care about is at
+	// the head of the finalizers list.
+	serviceClass, servicePlan, brokerName, brokerClient, err := c.getServiceClassPlanAndBroker(instance)
+	if err != nil {
+		return err
+	}
+
+	glog.V(4).Infof("Finalizing Instance %v/%v", instance.Namespace, instance.Name)
+
+	request := &brokerapi.DeleteServiceInstanceRequest{
+		ServiceID:         serviceClass.ExternalID,
+		PlanID:            servicePlan.ExternalID,
+		AcceptsIncomplete: true,
+	}
+
+	glog.V(4).Infof("Deprovisioning Instance %v/%v of ServiceClass %v at Broker %v", instance.Namespace, instance.Name, serviceClass.Name, brokerName)
+	response, respCode, err := brokerClient.DeleteServiceInstance(instance.Spec.ExternalID, request)
+
+	if err != nil {
+		s := fmt.Sprintf(
+			"Error deprovisioning Instance \"%s/%s\" of ServiceClass %q at Broker %q with status code %d: %s",
+			instance.Namespace,
+			instance.Name,
+			serviceClass.Name,
+			brokerName,
+			respCode,
+			err,
+		)
+		glog.Warning(s)
+		c.updateInstanceCondition(
+			instance,
+			v1alpha1.InstanceConditionReady,
+			v1alpha1.ConditionUnknown,
+			errorDeprovisionCalledReason,
+			"Deprovision call failed. "+s)
+		c.recorder.Event(instance, api.EventTypeWarning, errorDeprovisionCalledReason, s)
+		return err
+	}
+
+	if respCode == http.StatusAccepted {
+		glog.V(5).Infof("Received asynchronous de-provisioning response for Instance %v/%v of ServiceClass %v at Broker %v: response: %v", instance.Namespace, instance.Name, serviceClass.Name, brokerName, response)
+		if response.Operation != "" {
+			instance.Status.LastOperation = &response.Operation
+		}
+
+		// Tag this instance as having an ongoing async operation so we can enforce
+		// no other operations against it can start.
+		instance.Status.AsyncOpInProgress = true
+
+		err := c.updateInstanceCondition(
+			instance,
+			v1alpha1.InstanceConditionReady,
+			v1alpha1.ConditionFalse,
+			asyncDeprovisioningReason,
+			asyncDeprovisioningMessage,
+		)
+		if err != nil {
+			return err
+		}
+	} else {
+		err := c.updateInstanceCondition(
+			instance,
+			v1alpha1.InstanceConditionReady,
+			v1alpha1.ConditionFalse,
+			successDeprovisionReason,
+			successDeprovisionMessage,
```

Comment:
It's not necessarily a successful deprovision in this case - only if `http.StatusOK` is returned. Otherwise, we should update to a failure message. I've done something similar for provision in https://github.com/kubernetes-incubator/service-catalog/pull/923